### PR TITLE
Features/dao vacancies

### PR DESCRIPTION
**Pull request by 4en1x**

### CODE REVIEW

File: server/test/auth.test.js [hunk 1]
```diff
@@ -16,97 +16,102 @@ describe('#Autentification', () => {
...
 
   describe('#Get any page without authentication', () => {
-    it('should failed with 401 error', async () => {
-      const response = await req
-        .get(`${defaultUrl}/hirings`)
-        .set('Accept', 'application/json')
-        .ok(res => res.status <= 500);
-      expect(response.statusCode).to.equal(401);
-    });
+    it('This test should failed with 401 error because no one can see any page without authorization ',
```
Comment (aaneitchik) [inline server/test/auth.test.js]: should 'fail', not 'failed'

File: server/test/auth.test.js [hunk 2]
```diff
@@ -16,97 +16,102 @@ describe('#Autentification', () => {
...
-    it('should failed with 401 error : email doesn\'t exist', async () => {
-      const data = await readFileAsync('./test/data/auth/check-email-2.json', 'utf8');
-      const response = await req
-        .post(`${defaultUrl}/email`)
-        .send(JSON.parse(data))
-        .set('Accept', 'application/json')
-        .ok(res => res.status <= 500);
-      expect(response.statusCode).to.equal(401);
-    });
+    it('This testshould failed with 401 error : email doesn\'t exist',
```
Comment (aaneitchik) [inline server/test/auth.test.js]: test should fail

File: server/test/auth.test.js [hunk 3]
```diff
@@ -16,97 +16,102 @@ describe('#Autentification', () => {
...
+          .get(`${defaultUrl}/hirings`)
+          .set('Accept', 'application/json')
+          .ok(res => res.status <= 500);
+        expect(response.statusCode).to.equal(401);
+      });
   });
 
   describe('#Check email', () => {
-    it('should pass', async () => {
+    it('This test should pass, this email exist', async () => {
```
Comment (aaneitchik) [inline server/test/auth.test.js]: email exists

File: server/test/auth.test.js [hunk 4]
```diff
@@ -16,97 +16,102 @@ describe('#Autentification', () => {
...
+    it('This test should failed with 404 error, because page doesn\'t exist',
+      async () => {
+        const data = await readFileAsync('./test/data/auth/login-1.json', 'utf8');
+        let response = await req
+          .post(`${defaultUrl}/login`)
+          .set('Accept', 'application/json')
+          .send(JSON.parse(data));
+        expect(response.statusCode).to.equal(200);
+        response = await req
+          .get(`${defaultUrl}/hiringssss`)
```
Comment (aaneitchik) [inline server/test/auth.test.js]: hiringssss?

File: server/test/candidates.test.js [hunk 5]
```diff
@@ -22,156 +22,167 @@ describe('#Candidates-Api', () => {
...
+    it('This test should pass, because hr and admin have access for all candidates and can add them (another example with different fields)',
+      async () => {
+        const data = await readFileAsync('./test/data/candidates/add-candidate-5.json', 'utf8');
+        const response = await req
+          .post(`${defaultUrl}/candidates`)
+          .set('Accept', 'application/json')
+          .send(JSON.parse(data));
+        expect(response.statusCode).to.equal(200);
+      });
+    it('This test should failed with 500 error because city don\'t exist and admin can not add it',
```
Comment (aaneitchik) [inline server/test/candidates.test.js]: city doesn't exist

File: server/test/candidates.test.js [hunk 6]
```diff
@@ -22,156 +22,167 @@ describe('#Candidates-Api', () => {
...
+        const data = await readFileAsync('./test/data/candidates/add-candidate-2.json', 'utf8');
+        const response = await req
+          .post(`${defaultUrl}/candidates`)
+          .set('Accept', 'application/json')
+          .send(JSON.parse(data))
+          .ok(res => res.status <= 500);
+        expect(response.statusCode).to.equal(500);
+      });
+
+    it('This test should failed with 500 error because field don\'t exist and admin can not add it',
```
Comment (aaneitchik) [inline server/test/candidates.test.js]: field doesn't

File: server/test/candidates.test.js [hunk 7]
```diff
@@ -22,156 +22,167 @@ describe('#Candidates-Api', () => {
...
+      async () => {
+        const data = await readFileAsync('./test/data/candidates/check-candidate-1.json', 'utf8');
+        const response = await req
+          .get(`${defaultUrl}/candidates/1`)
+          .set('Accept', 'application/json');
+        expect(response.statusCode).to.equal(200);
+        expect(response.body).to.shallowDeepEqual(toCamel(JSON.parse(data)));
+      });
+
+    it('This test should failed with 500 error :id don\'t exist and we can not read it',
```
Comment (aaneitchik) [inline server/test/candidates.test.js]: :id doesn't

File: server/test/feedbacks.test.js [hunk 8]
```diff
@@ -28,77 +28,82 @@ describe('#Feedbacks-Api', () => {
...
-        .post(`${defaultUrl}/login`)
-        .send(JSON.parse(adminAuthData));
-      expect(response.statusCode).to.equal(200);
-      response = await req
-        .get(`${defaultUrl}/feedbacks/0`)
-        .set('Accept', 'application/json');
-      expect(response.statusCode).to.equal(200);
-      expect(response.body).to.shallowDeepEqual({ found: false });
-    });
+    it('This test should pass with text { found: false } - feedback don\'t exist',
```
Comment (aaneitchik) [inline server/test/feedbacks.test.js]: doesn't

File: server/test/feedbacks.test.js [hunk 9]
```diff
@@ -28,77 +28,82 @@ describe('#Feedbacks-Api', () => {
...
-        .post(`${defaultUrl}/login`)
-        .send(JSON.parse(userAuthData));
-      expect(response.statusCode).to.equal(200);
-      response = await req
-        .get(`${defaultUrl}/feedbacks/1`)
-        .set('Accept', 'application/json')
-        .ok(res => res.status <= 500);
-      expect(response.statusCode).to.equal(403);
-    });
+    it('This test should failed with 403 error: user have access only to his feedbacks, and it\' not him',
```
Comment (aaneitchik) [inline server/test/feedbacks.test.js]: user has ... and this one is not his

File: server/test/hirings.test.js [hunk 10]
```diff
@@ -18,155 +18,167 @@ describe('#Hirings-Api', () => {
...
-    it('should failed with 500 error : candidate is not exist', async () => {
-      const data = await readFileAsync('./test/data/hirings/add-hirings-1.json', 'utf8');
-      const response = await request
-        .post(`${defaultUrl}/hirings?candidate=41`)
-        .set('Accept', 'application/json')
-        .send(JSON.parse(data))
-        .ok(res => res.status <= 500);
-      expect(response.statusCode).to.equal(500);
-    });
+    it('This test should failed with 500 error because candidate is not exist. We can not appoint hiring.',
```
Comment (aaneitchik) [inline server/test/hirings.test.js]: does not exist

File: server/test/hirings.test.js [hunk 11]
```diff
@@ -18,155 +18,167 @@ describe('#Hirings-Api', () => {
...
+        expect(response.statusCode).to.equal(200);
+      });
 
-    it('should failed with 500 error: hiring is not exist', async () => {
-      const response = await request
-        .delete(`${defaultUrl}/hirings/685421`)
-        .ok(res => res.status <= 500);
-      expect(response.statusCode).to.equal(500);
-    });
+    it('This test should failed with 500 error because hiring is not exist and nobody can delete it',
```
Comment (aaneitchik) [inline server/test/hirings.test.js]: does not exist

File: server/test/hirings.test.js [hunk 12]
```diff
@@ -18,155 +18,167 @@ describe('#Hirings-Api', () => {
...
-      expect(response.statusCode).to.equal(200);
-      const data = await readFileAsync('./test/data/hirings/add-hirings-1.json', 'utf8');
-      response = await request
-        .post(`${defaultUrl}/hirings?candidate=7`)
-        .set('Accept', 'application/json')
-        .send(JSON.parse(data));
-      expect(response.statusCode).to.equal(200);
-      expect(response.body.added).to.equal(true);
-    });
+    it('This test should pass, because if we delete hiring, candidate now don\'t have hiring and we can appoint it',
```
Comment (aaneitchik) [inline server/test/hirings.test.js]: doesn't have hiring

File: server/test/interviews.test.js [hunk 13]
```diff
@@ -28,175 +28,188 @@ describe('#Interviews-Api', () => {
...
+          .post(`${defaultUrl}/login`)
+          .send(JSON.parse(userAuthData));
+        expect(response.statusCode).to.equal(200);
+        response = await req
+          .get(`${defaultUrl}/interviews/1`)
+          .set('Accept', 'application/json');
+        expect(response.statusCode).to.equal(200);
+      });
+
+    it('This test should failed with 403 error because user don\'t have access to this interview and can\'t get it - he not appointed to it',
```
Comment (aaneitchik) [inline server/test/interviews.test.js]: doesn't have access

File: server/test/interviews.test.js [hunk 14]
```diff
@@ -28,175 +28,188 @@ describe('#Interviews-Api', () => {
...
-        .post(`${defaultUrl}/login`)
-        .send(JSON.parse(adminAuthData));
-      expect(response.statusCode).to.equal(200);
-      response = await req
-        .delete(`${defaultUrl}/interviews/67`)
-        .set('Accept', 'application/json')
-        .ok(res => res.status <= 500);
-      expect(response.statusCode).to.equal(500);
-    });
+    it('This test should pass, because user is admin',
```
Comment (aaneitchik) [inline server/test/interviews.test.js]: What does this test do?

File: server/test/vacancies.test.js [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+        expect(response.statusCode).to.equal(200);
+
+        response = await req
+          .post(`${defaultUrl}/vacancies`)
+          .set('Accept', 'application/json')
+          .ok(res => res.status <= 500);
+        expect(response.statusCode).to.equal(403);
+      });
+
+    it('This test should pass - admin and hr have access to add vacancies. Also in this test I check that data is correctly saved in db',
```
Comment (aaneitchik) [inline server/test/vacancies.test.js]: "This test checks", not "I check"

File: server/dao/vacancies.dao.js [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+      await this.connection.commit();
+      return candidates;
+    } catch (err) {
+      return this.connection.rollback(() => {
+        throw err;
+      });
+    }
+  }
+
+  async update(id, vacancy, skills, cityName) {
```
Comment (ala-n) [inline server/dao/vacancies.dao.js]: Just notes for future refactoring (@LeanidShutau please take a notice):
If logic of update is just delete -> re-add with the same id, don't need to write all this unreadable content just call delete -> create.

File: server/dao/vacancies.dao.js [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        sql: `DELETE FROM vacancy_has_skills
+              WHERE vacancy_id = ?`,
+        values: [id],
+      });
+
+      await super.update(id, vacancy);
+
+      await Promise.all(skills.map(async (source, index) => {
+        const skillsId = await this.connection.queryAsync({
+          sql: `SELECT id FROM skills
```
Comment (ala-n) [inline server/dao/vacancies.dao.js]: + DAO skills expected.

File: server/dao/vacancies.dao.js [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+                WHERE name = ?`,
+          values: [skills[index].name],
+        });
+        await this.connection.queryAsync({
+          sql: `INSERT INTO vacancy_has_skills (skill_id, vacancy_id, weight)
```
Comment (ala-n) [inline server/dao/vacancies.dao.js]: It's can be common for VacancyDAO, just separate method.

**Pull request merged**